### DISCUSSION TITLE
Node.data / Edge.sourceHandle are undocumented

### DISCUSSION THREAD

**Discussion by ghost** [xyflow/xyflow#4052] [Q&A]
Both documentation and `nodes.d.ts` say that `Node.data` is of type `T = any` yet in the examples the default nodes are assigned with
`data: {label: 'some text'}` and the label is somehow drawn inside the Node. I assume that it is not of type any but of some other type that includes at least `label`  and maybe something else.

Same situation is with `DefaultEdge.type` is has type `string`, but it is not, because if `type` is assigned to `step`, the edge behavior changes.

It is sad that such core API has no proper documentation. How do I get all the available options for these interfaces?

**Comment by bcakmakoglu** [comment]
Node data *can be* of type `any` (though more strictly speaking it should probably be a `Record<string, any>`)
if you use a custom node type ReactFlow does not force you into any specific shape for the `data` property. 
`data.label` is only used for the built-in nodes, your nodes do not have to adhere to this schema at all
*you do not have to have `data: { label: 'Node 1' }` in your custom node data if you don't want to have that.*

`type` on Edges *can be* of type `string` - you are allowed to choose *any arbitrary string* as the edge type.
The types you refer to are, again, the built-in defaults that you can choose to use but aren't *forced to use*.
The same applies for `type` on Nodes.

However, in v12 `type` will be typed as not just `string` but either a `string` or one of the built-in types -> `type: 'smoothstep' | 'step' | 'straight' | 'default' | string`.

**Reply by ghost** [reply]
Thank you for the reply. 

V12 `type` update for `Edge` is what I was asking for. It would be also nice to have it documented at https://reactflow.dev/api-reference/types/edge

I perfectly understand that when designing my own Node I can define `data` as any type, but what if I use the default node? How do I know what data is accepted by it? Namely the `label` field - the only way to know about it as of time being is via exploring the examples or by trial and error. My suggestion is that it should be typed the same way as the plans for the `Edge.type` update: 
`DefaultData | CustomData` where `DefaultData` contains at least the `label` field and any other fields that it could use, if any. And `CustomData` is of type `Record<string, any>`

**Reply by bcakmakoglu** [reply]
Consider that you can overwrite the built-in node types (same as you assign custom nodes you can overwrite the built-in ones) and if the library *enforced* that the type of `data` *has to be* `{ label: string }` it wouldn't be accurate.
Tricky situation and I see where you're coming from but, imo, having the type be more accepting of more-or-less arbitrary data is preferable in this case since the library cannot know that you've overwritten the built-in type and possibly expect a completely different shape for the data property.

**Reply by ghost** [reply]
But doesn't the union type `DefaultData | CustomData` removes the enforcing and the same time documenting the native API?
Or if you are strongly against that, can the possible values of default `data` be at least documented at https://reactflow.dev/api-reference/types/node? Because I am using the library for several days and still don't know what else except `label` I can pass to the default node there?

**Reply by bcakmakoglu** [reply]
I'm not *strongly* against it - I am just trying to argue the rationale for the existing type definitions 😄
There can probably be some improvements done, though I'm not sure if that Union Type would solve this.
I agree that there should be some documentation on this in some form, true.

As for your question what data you can pass to default nodes - *anything* really.
You don't even have to pass a label if you don't want to, it's not a mandatory property of data for built-in nodes.

So you can do this:
```ts
{
	id: '1',
	position: { x: 0, y: 0 },
	data: { label: 'Node 1' },
}
```

or this
```ts
{
	id: '1',
	position: { x: 0, y: 0 },
	data: { foo: 'bar', baz: 'boo' }
}
```

or this
```ts
{
	id: '1',
	position: { x: 0, y: 0 },
	data: { label: 'Node 1', foo: 'bar', baz: 'boo' }
}
```

the `data` property is whatever you want it to be 😄

**Reply by ghost** [reply]
But if I pass `label` it is used by the default node to display the text passed inside the node, and `foo/bar` are not used for anything. All I want is that the values that are used by the default node are documented either by the TS type or at least reflected in the documentation. 

Same for the `Node.type` it can be any string value, from the TS perspective that is correct, but neither TS model, nor the documentation tells that `'input'` and `'output'` strings in addition to being a valid TS properties also control how the handles are displayed. (Also imo it is a bit weird that `type: 'input'` draws only the `source handle` and `type: 'output'` draws only the `target handle` - shouldn't that be vise versa?)

**Reply by bcakmakoglu** [reply]
I agree it could be documented that `{ label?: string }` is the data type for default nodes, although I'm still not quite sure if *typing* it explicitly that way would result in better DX for the previously mentioned reasons.

As to your question about input/output:
An *input* node represents an "argument", whereas an *output* node represents a "result".
Might be you think of it as "this node receives an input" and "this node receives an output"? 😅
